Refactor `ModDifficultyAdjust` to more elegantly track user override status

peppy opened the pull request:
As discussed in discord. I'll let the code speak for itself here, not much to add.

- Supersedes https://github.com/ppy/osu/pull/13814 (includes it but adjusts one or two expectations)
- Fixes https://github.com/ppy/osu/issues/13799

https://user-images.githubusercontent.com/191335/124893209-7feb6380-e015-11eb-844c-226257919407.mov

smoogipoo committed:
Add difficulty adjustment mod tests

peppy committed:
Enable tests and update expectations

peppy committed:
Initial storage changes

peppy committed:
Remove now unnecessary `BeatmapDifficulty.ID`

peppy committed:
Fix `ShowsDefaultIndicator` not actually being consumed

peppy committed:
Initial implementation of new flow (only working for approach rate)

peppy committed:
Add back extended limits support

peppy committed:
Avoid the need for per-settings control classes

peppy committed:
Use existing reflection methods to avoid manual binding of `ExtendedLimits`

peppy committed:
Add comprehensive tests of difficulty adjust settings

peppy committed:
Add more comments and xmldoc

peppy committed:
Simplify serialisation edge case by moving to `Value` override

peppy committed:
Rename control class to be more descriptive

bdach commented on osu.Game.Tests/Visual/UserInterface/TestSceneModDifficultyAdjustSettings.cs:
```diff
@@ -0,0 +1,205 @@
+            checkBindableAtValue("Circle Size", null);
+        }
+
+        [Test]
+        public void TestUserOverrideMaintainedOnMatchingBeatmapValue()
+        {
+            setBeatmapWithDifficultyParameters(2);
+
+            checkSliderAtValue("Circle Size", 2);
+            checkBindableAtValue("Circle Size", null);
+
+            setSliderValue("Circle Size", 9);
+            checkSliderAtValue("Circle Size", 9);
+            checkBindableAtValue("Circle Size", 9);
+
+            setBeatmapWithDifficultyParameters(4);
+
+            checkSliderAtValue("Circle Size", 9);
+            checkBindableAtValue("Circle Size", 9);
+        }
```
Bit of a discrepancy between the test name here and the test scenario. I *assume* `setSliderValue()` should have been called with 9 (or the second `setBeatmapWithDifficultyParameters()` with 4)?

peppy replied:
Yep, good catch.

bdach commented:
8 out of 9 files in this diff I like the look of a lot. The removal of these old hacks from `ModDifficultyAdjustment` is a sight for sore eyes.

The one I'm not super sure of is `DifficultyAdjustSettingsControl`. It starts to make sense after a bit of staring, but it's still a little complex (and maybe inevitably so, given that what it's doing is a little complex on its own). Maybe it can be improved with some renaming, not sure. I would definitely appreciate a second pair of eyes to look at it to make sure it's not just me.

peppy commented:
Definitely agree about that class being on the verge of too complicated. I'll probably do another pass myself tomorrow but I'm not sure how much it can be improved, as it is both fitting the existing `SettingsItem` mould and adding one extra dimension.

Even if it can't be heavily improved, at least it's now confined to this one class, though.

peppy committed:
Standardise variables

peppy committed:
Allow null values for `ReadCurrentFromDifficulty`

As long as this isn't a constructor parameter it feels best to
gracefully handle omission. Realistically having it in the ctor is the
best move, but it doesn't feel great in line with the other parameters
passed in via object initalisers.

peppy committed:
Update override matching test to match expectations

peppy committed:
Refactor `DifficultyAdjustSettingsControl` to help with readability

peppy committed:
Simplify bindable update methods

peppy commented:
@bdach give that another read now. i think the class should be a lot easier to mentally parse.

bdach committed:
Merge branch 'master' into da-mod-refactor

bdach committed:
Update test to match expectations after refactor

bdach reviewed (dismissed):
Yeah I like this as is. Not sure whether another pair of eyes is needed.

frenzibyte reviewed (changes requested):
Been randomly playing with the controls and noticed this:

https://user-images.githubusercontent.com/22781491/125178368-f2626c00-e1ec-11eb-8b52-d11fb039488a.mp4

Seems reproducible by first selecting then deslecting the mod, then activate again and change the values (w/o extended limits checked I think)

---

Also missing `SettingControlType` specification for the osu!catch variant, as "suggested" in the comments below.

peppy commented:
What is the video showing? I've watched 5 times and still have no idea.

frenzibyte commented:
"Approach rate" setting no longer showing the restore value indicator.

peppy commented:
Thanks. I remember there being some unbind code somewhere that may still be being triggered.

peppy committed:
Add missing osu!catch difficulty adjust attributes

bdach committed:
Do not store direct references to original bindable

`DifficultyAdjustSettingsControl` and its inner `SliderControl` were
holding different references to `DifficultyBindable`s from the
difficulty adjust mod, therefore leading to bindings being lost to the
framework-side automatic unbind logic if the mod was toggled off and
back on in rapid succession.

Resolve by adding a shadowed implementation of `GetBoundCopy()` and
using it to isolate the controls from the mod bindable.

bdach commented:
I believe I've resolved the issue. As far as I can tell this was caused by storing the `DifficultyBindable` raw in drawables (`DifficultyAdjustSettingsControl` and its inner `SliderControl`), which in turn caused the framework-side automatic unbinding code to fire and unbind the mod's `DifficultyBindable`s (as they were the same reference as the controls').

The fix is to add a specialised implementation of `GetBoundCopy()`. It's a bit dodgy around `MaxValue` and `ExtendedMaxValue` (assumes they're essentially constants); they can be changed to bindables if need be, but I figured the bare minimum might be enough for now.

smoogipoo commented on osu.Game/Rulesets/Mods/DifficultyBindable.cs:
```diff
@@ -0,0 +1,107 @@
+        /// An internal numeric bindable to hold and propagate min/max/precision.
+        /// The value of this bindable should not be set.
+        /// </summary>
+        public readonly BindableFloat CurrentNumber = new BindableFloat
```
If this bindable should not be set, then it should probably be `internal`?

smoogipoo reviewed (dismissed):
Looks okay

peppy committed:
Make `CurrentNumber` internal
